File: apps/demo-be-kotlin-ktor/src/test/kotlin/com/demobektkt/integration/steps/AttachmentSteps.kt
package com.demobektkt.integration.steps

import io.cucumber.java.en.Given
import io.cucumber.java.en.Then
import io.cucumber.java.en.When
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue

class AttachmentSteps {

    private fun uploadFile(
        username: String,
        filename: String,
        contentType: String,
        expenseIdKey: String,
        fileContent: ByteArray = "test file content".toByteArray(),
    ): Pair<Int, String> {
        val token = TestWorld.accessTokens[username] ?: error("$username has no access token")
        val expenseId =
            TestWorld.expenseIds[expenseIdKey] ?: error("no expense id stored at $expenseIdKey")
        return ServiceDispatcher.uploadAttachment(token, expenseId, filename, contentType, fileContent)
    }

    @Given(
        "bob has created an entry with body \\{ {string}: {string}, {string}: {string}, {string}: {string}, {string}: {string}, {string}: {string}, {string}: {string} \\}"
    )
    @Suppress("CyclomaticComplexMethod", "NestedBlockDepth")
    fun bobHasCreatedEntry(
        k1: String,
        v1: String,
        k2: String,
        v2: String,
        k3: String,
        v3: String,
        k4: String,
        v4: String,
        k5: String,
        v5: String,
        k6: String,
        v6: String,
    ) {
        // Ensure bob is logged in
        val passwords = listOf("Str0ng#Pass2", "Str0ng#Pass1")
        if (!TestWorld.accessTokens.containsKey("bob")) {
            for (pwd in passwords) {
                val (status, body) = ServiceDispatcher.login("bob", pwd)
                if (status == 200) {
                    JsonHelper.getString(body, "access_token")?.let {
                        TestWorld.accessTokens["bob"] = it
                    }
                    break
                }
            }
        }
        val token = TestWorld.accessTokens["bob"] ?: error("bob has no access token")
        val amount = v1.takeIf { k1 == "amount" }
            ?: v2.takeIf { k2 == "amount" }
            ?: v3.takeIf { k3 == "amount" }
            ?: v4.takeIf { k4 == "amount" }
            ?: v5.takeIf { k5 == "amount" }
            ?: v6.takeIf { k6 == "amount" }
            ?: error("amount key not found")
        val currency = v1.takeIf { k1 == "currency" }
            ?: v2.takeIf { k2 == "currency" }
            ?: v3.takeIf { k3 == "currency" }
            ?: v4.takeIf { k4 == "currency" }
            ?: v5.takeIf { k5 == "currency" }
            ?: v6.takeIf { k6 == "currency" }
            ?: error("currency key not found")
        val category = v1.takeIf { k1 == "category" }
            ?: v2.takeIf { k2 == "category" }
            ?: v3.takeIf { k3 == "category" }
            ?: v4.takeIf { k4 == "category" }
            ?: v5.takeIf { k5 == "category" }
            ?: v6.takeIf { k6 == "category" }
            ?: error("category key not found")
        val description = v1.takeIf { k1 == "description" }
            ?: v2.takeIf { k2 == "description" }
            ?: v3.takeIf { k3 == "description" }
            ?: v4.takeIf { k4 == "description" }
            ?: v5.takeIf { k5 == "description" }
            ?: v6.takeIf { k6 == "description" }
            ?: error("description key not found")
        val date = v1.takeIf { k1 == "date" }
            ?: v2.takeIf { k2 == "date" }
            ?: v3.takeIf { k3 == "date" }
            ?: v4.takeIf { k4 == "date" }
            ?: v5.takeIf { k5 == "date" }
            ?: v6.takeIf { k6 == "date" }
            ?: error("date key not found")
        val type = v1.takeIf { k1 == "type" }
            ?: v2.takeIf { k2 == "type" }
            ?: v3.takeIf { k3 == "type" }
            ?: v4.takeIf { k4 == "type" }
            ?: v5.takeIf { k5 == "type" }
            ?: v6.takeIf { k6 == "type" }
            ?: error("type key not found")

        val (status, respBody) =
            ServiceDispatcher.createExpense(
                token,
                amount,
                currency,
                category,
                description,
                date,
                type,
            )
        assertTrue(
            status == 201,
            "Expected 201 creating bob's expense, got $status. Body: $respBody",
        )
        JsonHelper.getString(respBody, "id")?.let { TestWorld.expenseIds["bob:last"] = it }
    }

    @Given("^alice has uploaded file \"([^\"]+)\" with content type \"([^\"]+)\" to the entry$")
    fun aliceHasUploadedFileToEntry(filename: String, contentType: String) {
        val (status, body) = uploadFile("alice", filename, contentType, "alice:last")
        assertTrue(status == 201, "Expected 201 uploading file, got $status. Body: $body")
        JsonHelper.getString(body, "id")?.let { TestWorld.attachmentIds["alice:last"] = it }
    }

    @When(
        "^alice uploads file \"([^\"]+)\" with content type \"([^\"]+)\" to POST /api/v1/expenses/\\{expenseId\\}/attachments$"
    )
    fun aliceUploadsFileToExpense(filename: String, contentType: String) {
        val (status, body) = uploadFile("alice", filename, contentType, "alice:last")
        TestWorld.lastResponseStatus = status
        TestWorld.lastResponseBody = body
        if (status == 201) {
            JsonHelper.getString(body, "id")?.let { TestWorld.attachmentIds["alice:last"] = it }
        }
    }

    @When(
        "^alice uploads file \"([^\"]+)\" with content type \"([^\"]+)\" to POST /api/v1/expenses/\\{bobExpenseId\\}/attachments$"
    )
    fun aliceUploadsToBobsExpense(filename: String, contentType: String) {
        val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
        val expenseId = TestWorld.expenseIds["bob:last"] ?: error("no bob expense id stored")
        val (status, body) =
            ServiceDispatcher.uploadAttachment(
                token,
                expenseId,
                filename,
                contentType,
                "test file content".toByteArray(),
            )
        TestWorld.lastResponseStatus = status
        TestWorld.lastResponseBody = body
    }

    @When(
        "alice uploads an oversized file to POST \\/api\\/v1\\/expenses\\/\\{expenseId\\}\\/attachments"
    )
    fun aliceUploadsOversizedFile() {
        val oversizedContent = ByteArray(11 * 1024 * 1024) { 'X'.code.toByte() } // 11 MB
        val (status, body) =
            uploadFile("alice", "large.jpg", "image/jpeg", "alice:last", oversizedContent)
        TestWorld.lastResponseStatus = status
        TestWorld.lastResponseBody = body
    }

    @When("alice sends GET \\/api\\/v1\\/expenses\\/\\{expenseId\\}\\/attachments")
    fun aliceSendsGetAttachments() {
        val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
        val expenseId = TestWorld.expenseIds["alice:last"] ?: error("no expense id stored")
        val (status, body) = ServiceDispatcher.listAttachments(token, expenseId)
        TestWorld.lastResponseStatus = status
        TestWorld.lastResponseBody = body
    }

    @When("^alice sends GET /api/v1/expenses/\\{bobExpenseId\\}/attachments$")
    fun aliceSendsGetBobsAttachments() {
        val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
        val expenseId = TestWorld.expenseIds["bob:last"] ?: error("no bob expense id stored")
        val (status, body) = ServiceDispatcher.listAttachments(token, expenseId)
        TestWorld.lastResponseStatus = status
        TestWorld.lastResponseBody = body
    }

    @When(
        "alice sends DELETE \\/api\\/v1\\/expenses\\/\\{expenseId\\}\\/attachments\\/\\{attachmentId\\}"
    )
    fun aliceSendsDeleteAttachment() {
        val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
        val expenseId = TestWorld.expenseIds["alice:last"] ?: error("no expense id stored")
        val attachmentId =
            TestWorld.attachmentIds["alice:last"] ?: error("no attachment id stored")
        val (status, body) = ServiceDispatcher.deleteAttachment(token, expenseId, attachmentId)
        TestWorld.lastResponseStatus = status
        TestWorld.lastResponseBody = body
    }

    @When("^alice sends DELETE /api/v1/expenses/\\{bobExpenseId\\}/attachments/\\{attachmentId\\}$")
    fun aliceSendsDeleteAttachmentOnBobsExpense() {
        val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
        val expenseId = TestWorld.expenseIds["bob:last"] ?: error("no bob expense id stored")
        val attachmentId =
            TestWorld.attachmentIds["alice:last"] ?: "00000000-0000-0000-0000-000000000000"
        val (status, body) = ServiceDispatcher.deleteAttachment(token, expenseId, attachmentId)
        TestWorld.lastResponseStatus = status
        TestWorld.lastResponseBody = body
    }

    @When(
        "^alice sends DELETE /api/v1/expenses/\\{expenseId\\}/attachments/\\{randomAttachmentId\\}$"
    )
    fun aliceSendsDeleteNonExistentAttachment() {
        val token = TestWorld.accessTokens["alice"] ?: error("alice has no access token")
        val expenseId = TestWorld.expenseIds["alice:last"] ?: error("no expense id stored")
        val randomId = "00000000-0000-0000-0000-000000000000"
        val (status, body) = ServiceDispatcher.deleteAttachment(token, expenseId, randomId)
        TestWorld.lastResponseStatus = status
        TestWorld.lastResponseBody = body
    }

    @Then("the response body should contain {int} items in the {string} array")
    fun theResponseBodyShouldContainItemsInArray(count: Int, arrayKey: String) {
        val actual = JsonHelper.getArraySize(TestWorld.lastResponseBody, arrayKey)
        assertEquals(
            count,
            actual,
            "Expected $count items in '$arrayKey' array in: ${TestWorld.lastResponseBody}",
        )
    }

    @Then("the response body should contain an attachment with {string} equal to {string}")
    fun theResponseBodyShouldContainAttachmentWithField(field: String, expected: String) {
        val body = TestWorld.lastResponseBody
        assertTrue(body.contains(expected), "Expected attachment with '$field'='$expected' in: $body")
    }
}

File: apps/demo-fs-ts-nextjs/test/unit/be-steps/attachment.steps.ts
import path from "path";
import { loadFeature, describeFeature } from "@amiceli/vitest-cucumber";
import { expect } from "vitest";
import { createTestContext, registerUser, loginUser, getAuth, type TestContext } from "./helpers/test-context";
import { MAX_ATTACHMENT_SIZE } from "@/lib/types";

const feature = await loadFeature(
  path.resolve(__dirname, "../../../../../specs/apps/demo/be/gherkin/expenses/attachments.feature"),
);

async function createExpense(ctx: TestContext, username: string, body: Record<string, unknown>): Promise<string> {
  const resp = await ctx.client.dispatch("POST", "/api/v1/expenses", body, getAuth(ctx, username));
  if (resp.status !== 201) throw new Error(`Failed: ${JSON.stringify(resp.body)}`);
  return (resp.body as { id: string }).id;
}

async function uploadFile(
  ctx: TestContext,
  username: string,
  expenseId: string,
  filename: string,
  contentType: string,
  data?: Buffer,
): Promise<{ status: number; body: unknown }> {
  const fileData = data ?? Buffer.from("fake file content for testing");
  const resp = await ctx.client.dispatch(
    "POST",
    `/api/v1/expenses/${expenseId}/attachments`,
    null,
    getAuth(ctx, username),
    { filename, contentType, size: fileData.length, data: fileData },
  );
  return resp;
}

describeFeature(feature, ({ Scenario, Background }) => {
  let ctx: TestContext;

  Background(({ Given, And }) => {
    Given("the API is running", () => {
      ctx = createTestContext();
    });

    And('a user "alice" is registered with email "alice@example.com" and password "Str0ng#Pass1"', async () => {
      await registerUser(ctx, "alice", "alice@example.com", "Str0ng#Pass1");
    });

    And('"alice" has logged in and stored the access token', async () => {
      await loginUser(ctx, "alice", "Str0ng#Pass1");
    });

    And(
      'alice has created an entry with body { "amount": "10.50", "currency": "USD", "category": "food", "description": "Lunch", "date": "2025-01-15", "type": "expense" }',
      async () => {
        ctx.context.expenseId = await createExpense(ctx, "alice", {
          amount: "10.50",
          currency: "USD",
          category: "food",
          description: "Lunch",
          date: "2025-01-15",
          type: "expense",
        });
      },
    );
  });

  Scenario("Upload JPEG image returns 201 with attachment metadata", ({ When, Then, And }) => {
    When(
      'alice uploads file "receipt.jpg" with content type "image/jpeg" to POST /api/v1/expenses/{expenseId}/attachments',
      async () => {
        ctx.response = await uploadFile(ctx, "alice", ctx.context.expenseId as string, "receipt.jpg", "image/jpeg");
        if (ctx.response.status === 201) {
          ctx.context.attachmentId = (ctx.response.body as { id: string }).id;
        }
      },
    );

    Then("the response status code should be 201", () => {
      expect(ctx.response!.status).toBe(201);
    });

    And('the response body should contain a non-null "id" field', () => {
      expect((ctx.response!.body as Record<string, unknown>).id).toBeDefined();
    });

    And('the response body should contain "filename" equal to "receipt.jpg"', () => {
      expect((ctx.response!.body as Record<string, unknown>).filename).toBe("receipt.jpg");
    });

    And('the response body should contain "contentType" equal to "image/jpeg"', () => {
      expect((ctx.response!.body as Record<string, unknown>).contentType).toBe("image/jpeg");
    });

    And('the response body should contain a non-null "url" field', () => {
      expect((ctx.response!.body as Record<string, unknown>).url).toBeDefined();
    });
  });

  Scenario("Upload PDF document returns 201 with attachment metadata", ({ When, Then, And }) => {
    When(
      'alice uploads file "invoice.pdf" with content type "application/pdf" to POST /api/v1/expenses/{expenseId}/attachments',
      async () => {
        ctx.response = await uploadFile(
          ctx,
          "alice",
          ctx.context.expenseId as string,
          "invoice.pdf",
          "application/pdf",
        );
      },
    );

    Then("the response status code should be 201", () => {
      expect(ctx.response!.status).toBe(201);
    });

    And('the response body should contain a non-null "id" field', () => {
      expect((ctx.response!.body as Record<string, unknown>).id).toBeDefined();
    });

    And('the response body should contain "filename" equal to "invoice.pdf"', () => {
      expect((ctx.response!.body as Record<string, unknown>).filename).toBe("invoice.pdf");
    });

    And('the response body should contain "contentType" equal to "application/pdf"', () => {
      expect((ctx.response!.body as Record<string, unknown>).contentType).toBe("application/pdf");
    });

    And('the response body should contain a non-null "url" field', () => {
      expect((ctx.response!.body as Record<string, unknown>).url).toBeDefined();
    });
  });

  Scenario("List attachments for an entry returns all uploaded files with metadata", ({ Given, And, When, Then }) => {
    Given('alice has uploaded file "receipt.jpg" with content type "image/jpeg" to the entry', async () => {
      const resp = await uploadFile(ctx, "alice", ctx.context.expenseId as string, "receipt.jpg", "image/jpeg");
      ctx.context.attachmentId = (resp.body as { id: string }).id;
    });

    And('alice has uploaded file "invoice.pdf" with content type "application/pdf" to the entry', async () => {
      await uploadFile(ctx, "alice", ctx.context.expenseId as string, "invoice.pdf", "application/pdf");
    });

    When("alice sends GET /api/v1/expenses/{expenseId}/attachments", async () => {
      ctx.response = await ctx.client.dispatch(
        "GET",
        `/api/v1/expenses/${ctx.context.expenseId}/attachments`,
        null,
        getAuth(ctx, "alice"),
      );
    });

    Then("the response status code should be 200", () => {
      expect(ctx.response!.status).toBe(200);
    });

    And('the response body should contain 2 items in the "attachments" array', () => {
      const body = ctx.response!.body as { attachments: unknown[] };
      expect(body.attachments).toHaveLength(2);
    });

    And('the response body should contain an attachment with "filename" equal to "receipt.jpg"', () => {
      const body = ctx.response!.body as { attachments: { filename: string }[] };
      expect(body.attachments.some((a) => a.filename === "receipt.jpg")).toBe(true);
    });

    And('the response body should contain an attachment with "filename" equal to "invoice.pdf"', () => {
      const body = ctx.response!.body as { attachments: { filename: string }[] };
      expect(body.attachments.some((a) => a.filename === "invoice.pdf")).toBe(true);
    });
  });

  Scenario("Delete attachment returns 204", ({ Given, When, Then }) => {
    Given('alice has uploaded file "receipt.jpg" with content type "image/jpeg" to the entry', async () => {
      const resp = await uploadFile(ctx, "alice", ctx.context.expenseId as string, "receipt.jpg", "image/jpeg");
      ctx.context.attachmentId = (resp.body as { id: string }).id;
    });

    When("alice sends DELETE /api/v1/expenses/{expenseId}/attachments/{attachmentId}", async () => {
      ctx.response = await ctx.client.dispatch(
        "DELETE",
        `/api/v1/expenses/${ctx.context.expenseId}/attachments/${ctx.context.attachmentId}`,
        null,
        getAuth(ctx, "alice"),
      );
    });

    Then("the response status code should be 204", () => {
      expect(ctx.response!.status).toBe(204);
    });
  });

  Scenario("Upload unsupported file type returns 415", ({ When, Then, And }) => {
    When(
      'alice uploads file "malware.exe" with content type "application/octet-stream" to POST /api/v1/expenses/{expenseId}/attachments',
      async () => {
        ctx.response = await uploadFile(
          ctx,
          "alice",
          ctx.context.expenseId as string,
          "malware.exe",
          "application/octet-stream",
        );
      },
    );

    Then("the response status code should be 415", () => {
      expect(ctx.response!.status).toBe(415);
    });

    And('the response body should contain a validation error for "file"', () => {
      const body = ctx.response!.body as Record<string, unknown>;
      expect(String(body.error).toLowerCase()).toContain("file");
    });
  });

  Scenario("Upload file exceeding the size limit returns 413", ({ When, Then, And }) => {
    When("alice uploads an oversized file to POST /api/v1/expenses/{expenseId}/attachments", async () => {
      const oversized = Buffer.alloc(MAX_ATTACHMENT_SIZE + 1, "x");
      ctx.response = await uploadFile(
        ctx,
        "alice",
        ctx.context.expenseId as string,
        "big.jpg",
        "image/jpeg",
        oversized,
      );
    });

    Then("the response status code should be 413", () => {
      expect(ctx.response!.status).toBe(413);
    });

    And("the response body should contain an error message about file size", () => {
      const body = ctx.response!.body as Record<string, unknown>;
      expect(String(body.error).toLowerCase()).toContain("size");
    });
  });

  Scenario("Upload attachment to another user's entry returns 403", ({ Given, And, When, Then }) => {
    Given('a user "bob" is registered with email "bob@example.com" and password "Str0ng#Pass2"', async () => {
      await registerUser(ctx, "bob", "bob@example.com", "Str0ng#Pass2");
      await loginUser(ctx, "bob", "Str0ng#Pass2");
    });

    And(
      'bob has created an entry with body { "amount": "25.00", "currency": "USD", "category": "transport", "description": "Taxi", "date": "2025-01-15", "type": "expense" }',
      async () => {
        ctx.context.bobExpenseId = await createExpense(ctx, "bob", {
          amount: "25.00",
          currency: "USD",
          category: "transport",
          description: "Taxi",
          date: "2025-01-15",
          type: "expense",
        });
      },
    );

    When(
      'alice uploads file "receipt.jpg" with content type "image/jpeg" to POST /api/v1/expenses/{bobExpenseId}/attachments',
      async () => {
        ctx.response = await uploadFile(ctx, "alice", ctx.context.bobExpenseId as string, "receipt.jpg", "image/jpeg");
      },
    );

    Then("the response status code should be 403", () => {
      expect(ctx.response!.status).toBe(403);
    });
  });

  Scenario("List attachments on another user's entry returns 403", ({ Given, And, When, Then }) => {
    Given('a user "bob" is registered with email "bob@example.com" and password "Str0ng#Pass2"', async () => {
      await registerUser(ctx, "bob", "bob@example.com", "Str0ng#Pass2");
      await loginUser(ctx, "bob", "Str0ng#Pass2");
    });

    And(
      'bob has created an entry with body { "amount": "25.00", "currency": "USD", "category": "transport", "description": "Taxi", "date": "2025-01-15", "type": "expense" }',
      async () => {
        ctx.context.bobExpenseId = await createExpense(ctx, "bob", {
          amount: "25.00",
          currency: "USD",
          category: "transport",
          description: "Taxi",
          date: "2025-01-15",
          type: "expense",
        });
      },
    );

    When("alice sends GET /api/v1/expenses/{bobExpenseId}/attachments", async () => {
      ctx.response = await ctx.client.dispatch(
        "GET",
        `/api/v1/expenses/${ctx.context.bobExpenseId}/attachments`,
        null,
        getAuth(ctx, "alice"),
      );
    });

    Then("the response status code should be 403", () => {
      expect(ctx.response!.status).toBe(403);
    });
  });

  Scenario("Delete attachment on another user's entry returns 403", ({ Given, And, When, Then }) => {
    Given('a user "bob" is registered with email "bob@example.com" and password "Str0ng#Pass2"', async () => {
      await registerUser(ctx, "bob", "bob@example.com", "Str0ng#Pass2");
      await loginUser(ctx, "bob", "Str0ng#Pass2");
    });

    And(
      'bob has created an entry with body { "amount": "25.00", "currency": "USD", "category": "transport", "description": "Taxi", "date": "2025-01-15", "type": "expense" }',
      async () => {
        ctx.context.bobExpenseId = await createExpense(ctx, "bob", {
          amount: "25.00",
          currency: "USD",
          category: "transport",
          description: "Taxi",
          date: "2025-01-15",
          type: "expense",
        });
      },
    );

    And('alice has uploaded file "receipt.jpg" with content type "image/jpeg" to the entry', async () => {
      // Upload to alice's own expense, not bob's
      const resp = await uploadFile(ctx, "alice", ctx.context.expenseId as string, "receipt.jpg", "image/jpeg");
      ctx.context.attachmentId = (resp.body as { id: string }).id;
    });

    When("alice sends DELETE /api/v1/expenses/{bobExpenseId}/attachments/{attachmentId}", async () => {
      ctx.response = await ctx.client.dispatch(
        "DELETE",
        `/api/v1/expenses/${ctx.context.bobExpenseId}/attachments/${ctx.context.attachmentId}`,
        null,
        getAuth(ctx, "alice"),
      );
    });

    Then("the response status code should be 403", () => {
      expect(ctx.response!.status).toBe(403);
    });
  });

  Scenario("Delete non-existent attachment returns 404", ({ Given, When, Then }) => {
    Given('alice has uploaded file "receipt.jpg" with content type "image/jpeg" to the entry', async () => {
      const resp = await uploadFile(ctx, "alice", ctx.context.expenseId as string, "receipt.jpg", "image/jpeg");
      ctx.context.attachmentId = (resp.body as { id: string }).id;
    });

    When("alice sends DELETE /api/v1/expenses/{expenseId}/attachments/{randomAttachmentId}", async () => {
      ctx.response = await ctx.client.dispatch(
        "DELETE",
        `/api/v1/expenses/${ctx.context.expenseId}/attachments/${crypto.randomUUID()}`,
        null,
        getAuth(ctx, "alice"),
      );
    });

    Then("the response status code should be 404", () => {
      expect(ctx.response!.status).toBe(404);
    });
  });
});
